File: lib/editor/trace/knoll-yliluoma.ts
/**
 * Knoll-Yliluoma "arbitrary-palette positional dithering algorithm" —
 * client mirror of `filter-service/app/knoll_yliluoma.py`.
 *
 * The algorithm replaces single-chip snapping with N-candidate sub-
 * sampling: each cell selects N palette chips whose RUNNING MEAN
 * approximates the cell's target colour, then a positional threshold
 * (blue-noise LUT) picks one of those N chips per cell. Adjacent cells
 * hitting different threshold bins emit different chips, so a uniform
 * target region renders as a pleasing N-chip mix — the spatial-
 * quantization property that classical "single-snap" misses (and
 * that Ulichney neighbour-invasion can't deliver on monochrome input,
 * because its decision depends on neighbour colours that don't yet
 * differ).
 *
 * Reference: Joel Yliluoma (2014), "Joel Yliluoma's arbitrary-palette
 * positional dithering algorithm" — the candidate-selection scheme is
 * Knoll's; the explicit positional sort + threshold mapping is
 * Yliluoma's contribution.
 *
 * Two pure pieces in this module:
 *   - `knollYliluomaCandidates` — given target + palette + N, returns
 *     the N candidate chip indices (with multiplicity, since the same
 *     chip may be picked twice when the target is close to it).
 *   - `thresholdBin` — given (x, y) + LUT + N, returns the candidate
 *     rank ∈ [0, N) to emit at that position.
 *
 * The full pipeline composition (loop over cells, sort candidates by
 * lightness, look up threshold bin, emit chip RGB) lives in PR-F so
 * the algorithm itself stays small + parity-testable in isolation.
 *
 * Distance metric: squared-Euclidean over the provided palette space.
 * Callers choose the space (OKLab from `lib/color/oklab.ts`, CIE Lab
 * from `lib/color/ciede2000.ts`, or any other linear-ish space) — the
 * candidate-mean target colour is computed in that same space, so
 * averaging works as long as the space is approximately linear in
 * perceptual mixing.
 */

/**
 * A 256×256 blue-noise LUT, flattened row-major (`lut[y * 256 + x]`).
 * Same binary layout the server uses (`blue_noise_256.bin`); the
 * `cell-texture.ts` loader serves it as a `Uint8Array` of length 65536.
 */
export type BlueNoiseLut = Uint8Array
export const BLUE_NOISE_LUT_SIZE = 256

/**
 * Pick `patternSize` palette indices whose running mean approximates
 * `target`.
 *
 * Algorithm (Yliluoma 2014, §2):
 *   At step `i` (1-indexed), the running mean after picking `c_i`
 *   will be `(sum_prev + palette[c_i]) / i`. We want that mean to
 *   approximate `target`, so we pick
 *         `c_i = argmin_j ‖palette[j] - (target·i - sum_prev)‖²`.
 *   The first pick (i=1) collapses to plain nearest-neighbour. Each
 *   subsequent pick corrects the residual error from the running
 *   mean — so the running mean asymptotically tracks `target` even
 *   when no single palette chip is close.
 *
 * @param target       target colour in the palette's space, length D
 * @param palette      (M × D) palette laid out row-major: `palette[i*D + d]`
 * @param paletteSize  number of palette chips (= M)
 * @param dim          feature dimension (= D, typically 3)
 * @param patternSize  N ≥ 1; number of candidates to pick. N=1 ≡ plain snap.
 * @returns            (N,) candidate indices (may repeat)
 */
export function knollYliluomaCandidates(
  target: ReadonlyArray<number>,
  palette: ReadonlyArray<number>,
  paletteSize: number,
  dim: number,
  patternSize: number,
): number[] {
  if (patternSize < 1) {
    throw new Error(`patternSize must be ≥ 1; got ${patternSize}`)
  }
  if (target.length !== dim) {
    throw new Error(`target length ${target.length} doesn't match dim ${dim}`)
  }
  if (palette.length !== paletteSize * dim) {
    throw new Error(
      `palette length ${palette.length} doesn't match paletteSize×dim = ${paletteSize * dim}`,
    )
  }

  const candidates: number[] = new Array(patternSize)
  const cumulative = new Float64Array(dim) // zeros
  const residual = new Float64Array(dim)

  for (let i = 1; i <= patternSize; i += 1) {
    for (let d = 0; d < dim; d += 1) {
      residual[d] = target[d] * i - cumulative[d]
    }
    let bestJ = 0
    let bestD = Infinity
    for (let j = 0; j < paletteSize; j += 1) {
      const base = j * dim
      let s = 0
      for (let d = 0; d < dim; d += 1) {
        const diff = palette[base + d] - residual[d]
        s += diff * diff
      }
      if (s < bestD) {
        bestD = s
        bestJ = j
      }
    }
    candidates[i - 1] = bestJ
    const baseChosen = bestJ * dim
    for (let d = 0; d < dim; d += 1) {
      cumulative[d] += palette[baseChosen + d]
    }
  }
  return candidates
}

/**
 * Position → candidate rank ∈ `[0, patternSize)`.
 *
 * The 256×256 blue-noise LUT distributes values 0..255 organically
 * (no banding, no clusters per Ulichney 1993 void-and-cluster).
 * Mapping into N equal-width bins gives a sequence of N-tone
 * thresholds that look pleasant on uniform fields.
 *
 * Tile via `% 256` so positions outside the LUT wrap deterministically.
 */
export function thresholdBin(
  x: number,
  y: number,
  patternSize: number,
  lut: BlueNoiseLut,
): number {
  if (patternSize < 1) {
    throw new Error(`patternSize must be ≥ 1; got ${patternSize}`)
  }
  if (lut.length !== BLUE_NOISE_LUT_SIZE * BLUE_NOISE_LUT_SIZE) {
    throw new Error(
      `LUT length ${lut.length} != ${BLUE_NOISE_LUT_SIZE}² — wrong binary?`,
    )
  }
  const wrappedX = ((x % BLUE_NOISE_LUT_SIZE) + BLUE_NOISE_LUT_SIZE) % BLUE_NOISE_LUT_SIZE
  const wrappedY = ((y % BLUE_NOISE_LUT_SIZE) + BLUE_NOISE_LUT_SIZE) % BLUE_NOISE_LUT_SIZE
  const raw = lut[wrappedY * BLUE_NOISE_LUT_SIZE + wrappedX]
  return Math.floor((raw * patternSize) / 256)
}

/**
 * Stable sort of candidate indices by `palette[idx, axis]`.
 *
 * Yliluoma's variant sorts by lightness so low-threshold positions
 * (dark blue-noise values) get the darkest candidate and vice-versa
 * — this makes the dither pattern look like graceful tone-mapping
 * rather than random noise. `axis = 0` matches OKLab/CIE Lab where
 * the first component is L.
 *
 * Stable so repeated candidates keep their first-pick order, matching
 * numpy's stable argsort.
 */
export function candidatesSortedByAxis(
  candidates: ReadonlyArray<number>,
  palette: ReadonlyArray<number>,
  dim: number,
  axis = 0,
): number[] {
  return candidates
    .map((idx, i) => ({ idx, i, key: palette[idx * dim + axis] }))
    .sort((a, b) => {
      if (a.key !== b.key) return a.key - b.key
      return a.i - b.i // stable: preserve insertion order on ties
    })
    .map((x) => x.idx)
}

File: filter-service/app/knoll_yliluoma.py
"""
Knoll-Yliluoma "arbitrary-palette positional dithering algorithm"
(Yliluoma 2014) — candidate selection + threshold-position lookup.

The algorithm replaces single-chip snapping with N-candidate sub-
sampling: each cell selects N palette chips whose RUNNING MEAN
approximates the cell's target colour, then a positional threshold
(blue-noise LUT) picks one of those N chips per cell. Adjacent cells
hitting different threshold bins emit different chips, so a uniform
target region renders as a pleasing N-chip mix — the spatial-
quantization property that classical "single-snap" misses (and that
Ulichney neighbour-invasion can't deliver on monochrome input,
because its decision depends on neighbour colours that don't yet
differ).

Reference: Joel Yliluoma (2014), "Joel Yliluoma's arbitrary-palette
positional dithering algorithm" — the candidate-selection scheme is
Knoll's; the explicit positional sort + threshold mapping is
Yliluoma's contribution.

Two pure pieces in this module:
  - `knoll_yliluoma_candidates` — given target + palette + N, returns
    the N candidate chip indices (with multiplicity, since the same
    chip may be picked twice when the target is close to it).
  - `threshold_bin` — given (x, y) + LUT + N, returns the candidate
    rank ∈ [0, N) to emit at that position.

The full pipeline composition (loop over cells, sort candidates by
lightness, look up threshold bin, emit chip RGB) lives in PR-F so the
algorithm itself stays small + parity-testable in isolation.

A `lib/editor/trace/knoll-yliluoma.ts` mirror runs in the client
preview; parity is asserted by the same constructed test cases.

Distance metric: the algorithm uses squared-Euclidean over the
provided palette space. Callers choose the space (OKLab from
`oklab.py`, CIE Lab from `ciede2000.py`, or any other linear-ish
space) — the candidate-mean target colour is computed in that same
space, so averaging works as long as the space is approximately
linear in perceptual mixing.
"""
from __future__ import annotations

from pathlib import Path

import numpy as np

# Reuse the same 256×256 blue-noise LUT already used by `cell_texture.py`
# so the dithering pattern is consistent with the existing texture step.
# Same binary is served to the browser via `public/assets/blue-noise-256.bin`
# so client previews use byte-identical thresholds.
_LUT_PATH = Path(__file__).parent / "data" / "blue_noise_256.bin"
BLUE_NOISE_LUT: np.ndarray = np.fromfile(_LUT_PATH, dtype=np.uint8).reshape(256, 256)


def knoll_yliluoma_candidates(
    target: np.ndarray,
    palette: np.ndarray,
    pattern_size: int,
) -> np.ndarray:
    """Pick `pattern_size` palette indices whose running mean
    approximates `target`.

    Algorithm (Yliluoma 2014, §2):
      At step `i` (1-indexed), the running mean after picking `c_i`
      will be `(sum_prev + palette[c_i]) / i`. We want that mean to
      approximate `target`, so we pick
            `c_i = argmin_j ‖palette[j] - (target·i - sum_prev)‖²`.
      That residual target `target·i - sum_prev` is exactly the chip
      whose addition makes the running mean coincide with `target`.

    The first pick (i=1) collapses to plain nearest-neighbour: the
    residual target equals `target` itself with sum_prev = 0. Each
    subsequent pick corrects the residual error from the running
    mean — so the running mean asymptotically tracks `target` even
    when no single palette chip is close.

    Args:
      target:        (3,) target colour in the palette's space.
      palette:       (M, 3) palette in the same space.
      pattern_size:  N ≥ 1; the number of candidates to pick. N=1
                     degenerates to plain nearest-neighbour.

    Returns:
      (N,) int64 indices into `palette` (may repeat — the same chip
      can be picked multiple times when target is close to it).
    """
    if pattern_size < 1:
        raise ValueError(f"pattern_size must be ≥ 1; got {pattern_size}")
    target = np.asarray(target, dtype=np.float64)
    palette = np.asarray(palette, dtype=np.float64)
    if target.shape != (palette.shape[1],):
        raise ValueError(
            f"target shape {target.shape} doesn't match palette feature dim {palette.shape[1]}"
        )

    candidates = np.empty(pattern_size, dtype=np.int64)
    cumulative_sum = np.zeros_like(target)
    for i in range(1, pattern_size + 1):
        residual_target = target * i - cumulative_sum
        d = ((palette - residual_target) ** 2).sum(axis=1)
        chosen = int(np.argmin(d))
        candidates[i - 1] = chosen
        cumulative_sum += palette[chosen]
    return candidates


def threshold_bin(
    x: int, y: int, pattern_size: int, lut: np.ndarray = BLUE_NOISE_LUT
) -> int:
    """Position → candidate rank ∈ `[0, pattern_size)`.

    The 256×256 blue-noise LUT distributes values 0..255 organically
    (no banding, no clusters per Ulichney 1993 void-and-cluster).
    Mapping into N equal-width bins gives a sequence of N-tone
    thresholds that look pleasant on uniform fields.

    Tile via `% 256` so positions outside the LUT wrap deterministically
    — the same `(x, y)` always lands on the same threshold value
    regardless of image size.
    """
    if pattern_size < 1:
        raise ValueError(f"pattern_size must be ≥ 1; got {pattern_size}")
    raw = int(lut[y % 256, x % 256])
    return (raw * pattern_size) // 256


def candidates_sorted_by_axis(
    candidates: np.ndarray, palette: np.ndarray, axis: int = 0
) -> np.ndarray:
    """Stable sort of candidate indices by `palette[idx, axis]`.

    Yliluoma's variant sorts by lightness so low-threshold positions
    (dark blue-noise values) get the darkest candidate and vice-versa
    — this makes the dither pattern look like graceful tone-mapping
    rather than random noise. `axis = 0` matches OKLab/CIE Lab where
    the first component is L.

    Stable so repeated candidates keep their first-pick order, which
    matters for parity between languages (numpy and Array.sort agree
    only on stable orderings).
    """
    palette = np.asarray(palette, dtype=np.float64)
    keys = palette[candidates, axis]
    order = np.argsort(keys, kind="stable")
    return candidates[order]
